parlai/utils/distributed.py

```diff
@@ -229,21 +230,27 @@ def sync_object(data, max_size=16384):
     return data
 
 
-def check_synced_parameters(model):
+def sync_parameters(model: torch.nn.Module) -> bool:
     """
-    Check that all parameters across all workers are the same.
+    Sync all parameters across all workers are the same.
 
-    Always returns True, or raises an AssertionError if they are not
-    synchronized.
+    Always returns True, or raises an AssertionError if there was a failure.
 
-    :param torch.nn.Module model: A pytorch model.
-    :return: True
+    :param model: A pytorch model.
+    :return: always True
     """
     if not is_distributed():
         # if things aren't distributed, of course things are in sync
         return True
 
-    # compute the local norm:
+    # syncrhonize all the parameters
+    with torch.no_grad():
+        for p in model.parameters():
+            if not is_primary_worker():
+                p.data.zero_()
+            dist.all_reduce(p.data)
```

Comment:
does all_reduce need to be called with primary worker or is it ok for any of the workers to call it? seems like this would make all params 0?

Reply:
all_reduce sums the result across all workers, including the primary worker. Hence we want to zero everyone out, EXCEPT the primary worker, and then take the sum. The result will be everyone has the primary worker's parameters.

Reply:
ok, thank you, didn't realize the implicit op was sum but that makes sense

Reply:
it has an optional reduce arg which I'll add for clarity